src/client/conn.rs

```diff
@@ -44,70 +45,63 @@ pub trait Connection {
 pub struct NetConnection {
     host: Mutex<String>,
     port: Mutex<u16>,
-    reader: Mutex<NetReader>,
-    writer: Mutex<NetWriter>,
+    stream: Mutex<NetBufStream>,
 }
 
 impl NetConnection {
-    fn new(host: &str, port: u16, reader: NetReader, writer: NetWriter) -> NetConnection {
+    fn new(host: &str, port: u16, stream: NetBufStream) -> NetConnection {
         NetConnection {
             host: Mutex::new(host.to_owned()),
             port: Mutex::new(port),
-            reader: Mutex::new(reader),
-            writer: Mutex::new(writer),
+            stream: Mutex::new(stream),
         }
     }
 
     /// Creates a thread-safe TCP connection to the specified server.
     pub fn connect(host: &str, port: u16) -> Result<NetConnection> {
-        let (reader, writer) = try!(NetConnection::connect_internal(host, port));
-        Ok(NetConnection::new(host, port, reader, writer))
+        let stream = try!(NetConnection::connect_internal(host, port));
+        Ok(NetConnection::new(host, port, stream))
     }
 
     /// connects to the specified server and returns a reader-writer pair.
-    fn connect_internal(host: &str, port: u16) -> Result<NetReadWritePair> {
+    fn connect_internal(host: &str, port: u16) -> Result<NetBufStream> {
         let socket = try!(TcpStream::connect(&format!("{}:{}", host, port)[..]));
-        Ok((BufReader::new(NetStream::Unsecured(try!(socket.try_clone()))),
-            BufWriter::new(NetStream::Unsecured(socket))))
+        Ok(BufStream::new(NetStream::Unsecured(socket)))
     }
 
     /// Creates a thread-safe TCP connection to the specified server over SSL.
     /// If the library is compiled without SSL support, this method panics.
     pub fn connect_ssl(host: &str, port: u16) -> Result<NetConnection> {
-        let (reader, writer) = try!(NetConnection::connect_ssl_internal(host, port));
-        Ok(NetConnection::new(host, port, reader, writer))
+        let stream = try!(NetConnection::connect_ssl_internal(host, port));
+        Ok(NetConnection::new(host, port, stream))
     }
 
     /// Connects over SSL to the specified server and returns a reader-writer pair.
     #[cfg(feature = "ssl")]
-    fn connect_ssl_internal(host: &str, port: u16) -> Result<NetReadWritePair> {
-        let socket = try!(TcpStream::connect(&format!("{}:{}", host, port)[..]));
-        let ssl = try!(ssl_to_io(SslContext::new(SslMethod::Sslv23)));
-        let ssl_socket = try!(ssl_to_io(SslStream::connect_generic(&ssl, socket)));
-        Ok((BufReader::new(NetStream::Ssl(try!(ssl_socket.try_clone()))),
-            BufWriter::new(NetStream::Ssl(ssl_socket))))
+    fn connect_ssl_internal(host: &str, port: u16) -> Result<NetBufStream> {
+        let domain = format!("{}:{}", host, port);
+        let socket = try!(TcpStream::connect(&domain));
+        let connector = try!(ssl_to_io(SslConnectorBuilder::new(SslMethod::tls()).map(|c| c.build()).map_err(Into::into)));
```

Comment:
There should probably be the option of configuring the connector here - people may want to e.g. add trusted root certificates.